optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/cloner/gizmo/GizmoSolutionOrEntityDescriptor.java

```diff
@@ -0,0 +1,117 @@
+
+        for (Field field : solutionFieldToMemberDescriptorMap.keySet()) {
+            if (DeepCloningUtils.getDeepCloneDecision(solutionDescriptor, field, solutionDescriptor.getSolutionClass(),
+                    field.getType())) {
+                deepClonedFields.add(field);
+            } else {
+                shallowlyClonedFields.add(field);
+            }
+        }
+    }
+
+    public GizmoSolutionOrEntityDescriptor(SolutionDescriptor<?> solutionDescriptor, Class<?> entityClass) {
+        this.solutionDescriptor = solutionDescriptor;
+        solutionFieldToMemberDescriptorMap = new HashMap<>();
+        addFieldsToSolutionFieldToMemberDescriptorMap(entityClass, solutionFieldToMemberDescriptorMap);
+        deepClonedFields = new HashSet<>();
+        shallowlyClonedFields = new HashSet<>();
+
+        for (Field field : solutionFieldToMemberDescriptorMap.keySet()) {
+            if (DeepCloningUtils.getDeepCloneDecision(solutionDescriptor, field, entityClass, field.getType())) {
+                deepClonedFields.add(field);
+            } else {
+                shallowlyClonedFields.add(field);
+            }
+        }
+    }
+
+    public GizmoSolutionOrEntityDescriptor(Class<?> holderClass, SolutionDescriptor<?> solutionDescriptor,
+            Map<Field, GizmoMemberDescriptor> solutionFieldToMemberDescriptorMap) {
+        this.solutionDescriptor = solutionDescriptor;
+        this.solutionFieldToMemberDescriptorMap = solutionFieldToMemberDescriptorMap;
+        deepClonedFields = new HashSet<>();
+        shallowlyClonedFields = new HashSet<>();
+
+        for (Field field : solutionFieldToMemberDescriptorMap.keySet()) {
+            if (DeepCloningUtils.getDeepCloneDecision(solutionDescriptor, field, holderClass, field.getType())) {
+                deepClonedFields.add(field);
+            } else {
+                shallowlyClonedFields.add(field);
+            }
+        }
+    }
+
+    private static void addFieldsToSolutionFieldToMemberDescriptorMap(Class<?> clazz,
+            Map<Field, GizmoMemberDescriptor> solutionFieldToMemberDescriptorMap) {
+        for (Field field : clazz.getDeclaredFields()) {
+            if (!Modifier.isStatic(field.getModifiers())) {
+                solutionFieldToMemberDescriptorMap.put(field, new GizmoMemberDescriptor(field));
+            }
+        }
+        if (clazz.getSuperclass() != null) {
+            addFieldsToSolutionFieldToMemberDescriptorMap(clazz.getSuperclass(), solutionFieldToMemberDescriptorMap);
+        }
+    }
+
+    public SolutionDescriptor<?> getSolutionDescriptor() {
+        return solutionDescriptor;
+    }
+
+    public Set<GizmoMemberDescriptor> getShallowClonedMemberDescriptors() {
+        return solutionFieldToMemberDescriptorMap.keySet().stream()
+                .filter(field -> shallowlyClonedFields.contains(field))
+                .map(solutionFieldToMemberDescriptorMap::get).collect(Collectors.toSet());
+    }
```

Comment:
This is a whole lot of calculation that should really be cached if it ends up on the hot path.
I assume you've made sure it isn't used on the hot path?

Reply:
This code is only called once (at build time for Quarkus, on creating the Solver in plain Java). The code is called once per GizmoSolutionOrEntityDescriptor to determine the fields to shallowly clone (i.e. each MemberDescriptor is only used once).